### ISSUE TITLE
Nondeterministic service startup failure

### ISSUE THREAD

**Issue by mtornwall** [Starcounter/Starcounter#2962]
I have set the Starcounter service to start automatically on my Windows Server 2012 R2 VM. For some reason the server sometimes fails to start, both during system startup and when manually started, with the following error message.

```
Service cannot be started. System.Exception: ScErrDatabaseEngineTerminated (SCERR10007): The server detected that the user code host executable (sccode.exe) unexpectedly terminated during a management operation.
Version: 2.1.57.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10007.
   at Starcounter.Server.ServerServiceProcess.WaitUntilServerIsOnlineOrSignalExit[T](Func`2 checkExited, T context) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Server\Starcounter.Server\ServerServiceProcess.cs:line 217
   at Starcounter.Server.ServerServiceProcess.OnStart(String[] args) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Server\Starcounter.Server\ServerServiceProcess.cs:line 151
   at System.ServiceProcess.ServiceBase.ServiceQueuedMainCallback(Object state)
```

I have seen the same behavior on my Windows 7 development machine and on a Windows Server 2016 Technical Preview VM.

I'm happy to demonstrate this problem in a live environment if it would help.

**Comment by dan31** [comment]
CC @per-samuelsson Maybe you can take a guess here along with other procmanagement-related things you are about to take.

**Comment by per-samuelsson** [comment]
@mtornwall 

Sorry you're experiencing this. Not easy to guess what's causing it given just this info. But I guess it happens quite frequently for you, since you can confidently demonstrate it?

A wild guess is that it's something caused by multiple startups at the same time. For instance, that the server, running as a Windows service, starts in parallell to a on-demand started instance of the same service (sound confusing, right? :smile:).

Anyway,

1. Something in the logs? (`staradmin list --max=20 log notice | clip` to put the last 20 logs, severity notice and up, on the clipboard. Paste them here).
2. Did this start recently? We've done some things with service starts in the installer recently (not sure it's pushed though). Might be caused by that?

**Comment by mtornwall** [comment]
Curiously this seems to happen much less frequently once the service has successfully started at least once. This may be a coincidence, but I'll look into it some more.

> Did this start recently? We've done some things with service starts in the installer recently (not sure it's pushed though). Might be caused by that?

Unfortunately I cannot say as I've only tried running the service with the 2.1.x versions.

**Comment by mtornwall** [comment]
OK, I managed to observe this on my production VM again. The log contained mostly application errors that I know the causes for, but here are some items from the failed start this morning:

```
[08:43:26, Critical, Starcounter (administrator)]
System.Exception: ScErrUnspecified (SCERR999): An unspecified error caused the operation to fail. System.Net.Sockets.SocketException (0x80004005): No connection could be made because the target machine actively refused it [::ffff:127.0.0.1]:8181
   at System.Net.Sockets.Socket.Connect(IPAddress[] addresses, Int32 port)
   at System.Net.Sockets.Socket.Connect(String host, Int32 port)
   at Starcounter.NodeTask.PerformSyncRequest() in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter.Internal\Rest\NodeTask.cs:line 532
Version: 2.1.57.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR999.
   at Starcounter.GatewayHandlers.RegisterUriHandlerNative(UInt16 port, String appName, String originalUriInfo, String processedUriInfo, Byte[] nativeParamTypes, UInt16 managedHandlerIndex, UInt64& handlerInfo) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter\Apps\GatewayHandlers.cs:line 458
   at Starcounter.Rest.UriHandlersManager.RegisterUriHandler(UInt16 port, String originalUriInfo, String processedUriInfo, Byte[] nativeParamTypes, Type messageType, Func`4 wrappedDelegate, NetworkProtocolType protoType, HandlerOptions ho) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 672
   at Starcounter.Rest.UriManagedHandlersCodegen.RegisterDelegate(UInt16 port, String methodSpaceUri, MethodInfo userDelegateInfo, Expression delegExpr, NetworkProtocolType protoType, HandlerOptions ho) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 614
   at Starcounter.Rest.RestRegistrationProxy.RegisterHandler[T1,T2](UInt16 port, String methodSpaceUri, Func`3 handler, HandlerOptions ho) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter.Rest\RestInitialization.cs:line 78
   at Starcounter.Internal.PuppetRestHandler.RegisterJsonPatchHandlers(UInt16 port) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\PuppetRestHandler.cs:line 115
   at Starcounter.Internal.AppsBootstrapper.InitAppsBootstrapper(Byte numSchedulers, UInt16 defaultUserHttpPort, UInt16 defaultSystemHttpPort, UInt32 sessionTimeoutMinutes, String dbName, Boolean noNetworkGateway) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:line 110
   at StarcounterInternal.Bootstrap.Control.Setup(String[] args) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter.Bootstrap\Control.cs:line 204
   at StarcounterInternal.Bootstrap.Control.Main(String[] args) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-26\Level1\src\Starcounter.Bootstrap\Control.cs:line 55
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR999


[08:43:27, Error, Starcounter (service)]
ScErrDatabaseEngineTerminated (SCERR10007): The server detected that the user code host executable (sccode.exe) unexpectedly terminated during a management operation.

[08:43:27, Error, Starcounter (service)]
ScErrUnspecified (SCERR999): An unspecified error caused the operation to fail.
```

**Comment by per-samuelsson** [comment]
> The log contained mostly application errors that I know the causes for, but here are some items from the failed start this morning:

Thanks @mtornwall! That can be of great help I think. 

@alemoi, do you have a take on that?

**Comment by alemoi** [comment]
> @alemoi, do you have a take on that?

Sure! It seems that `scadminserver` Administrator process is starting faster than gateway. I will add a step controller in scservice!

**Comment by alemoi** [comment]
Big thanks for reporting this. The issue should be fixed via 7742306a700fef3a867661d1069e4519cc75ab5e
Available in the following nightly. Would be nice if @mtornwall would confirm the fix! Thank you!

**Comment by dan31** [comment]
@alemoi, so where is our State Machine for the processes? :)

On Wednesday, 21 October 2015, Alexey <notifications@github.com> wrote:

> Big thanks for reporting this. The issue should be fixed via 7742306
> <https://github.com/Starcounter/Starcounter/commit/7742306a700fef3a867661d1069e4519cc75ab5e>
> Available in the following nightly. Would be nice if @mtornwall
> <https://github.com/mtornwall> would confirm the fix! Thank you!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Starcounter/Starcounter/issues/2962#issuecomment-149916940>
> .
>


-- 
Kind regards,
Dan Skatov

**Comment by per-samuelsson** [comment]
> @alemoi, so where is our State Machine for the processes? :)

Haha! Is that a rhetorical question? Considering **you** are the state machine. But @alemoi is also of course. We all are. Everything is. :wink: 

Nice you solved this one, @alemoi :+1:

**Comment by mtornwall** [comment]
@alemoi Glad I could help. :-) I conducted a quick test on my Windows Server 2016 VM and got five out of five successful service startups. I will define that as "working" until proven wrong. :-)

I'll close the issue now. In the off chance that this bug is in fact still present I will reopen it.

**Comment by mtornwall** [comment]
Now it happened again, this time on version 2.1.257 (the most recent nightly as of this writing). This time the Starcounter log simply said:

```
[18:41:40, Error, Starcounter (service)]
ScErrStartNetworkGateway (SCERR2110): Problems starting network gateway process.
```

The following message was left in the system log:

```
Service cannot be started. System.Exception: ScErrStartNetworkGateway (SCERR2110): Problems starting network gateway process.
Version: 2.1.257.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR2110.
   at Starcounter.Server.ServerServiceProcess.WaitUntilServerIsOnlineOrSignalExit[T](Func`2 checkExited, T context) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-53\Level1\src\Server\Starcounter.Server\ServerServiceProcess.cs:line 217
   at Starcounter.Server.ServerServiceProcess.OnStart(String[] args) in c:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-53\Level1\src\Server\Starcounter.Server\ServerServiceProcess.cs:line 151
   at System.ServiceProcess.ServiceBase.ServiceQueuedMainCallback(Object state)
```

@alemoi

**Comment by alemoi** [comment]
Thank you for reporting! It even might be a different problem! I will go through start up sequence and put verifications there.

**Comment by alemoi** [comment]
I just produced a special build of SC that logs gateway startup sequence. @mtornwall can you please install it and post here the last log contents when you experience a stuck? Thank you!
Here is the setup: http://downloads.starcounter.com/download/Starcounter/DevelopDailyBuilds/2.1.260

**Comment by mtornwall** [comment]
Thanks for the characteristically rapid response @alemoi. :-) I will set this up on my testing server tonight.

**Comment by alemoi** [comment]
@mtornwall was you able to reproduce this? Thanks!

**Comment by mtornwall** [comment]
@alemoi It's been running for a few days now, but unfortunately I cannot simply reboot that machine at will. I can try and set it up in another environment and try to provoke it into breaking.

**Comment by alemoi** [comment]
Thank you very much!

**Comment by dan31** [comment]
This one is important, invokes unintended downtime. Maybe @alemoi we'd have to address it more deliberately, such as running this pattern in an acceptance-test like way.

**Comment by alemoi** [comment]
Agreed, but let's wait until this is reproduced again.

**Comment by per-samuelsson** [comment]
What strategy do we have now when there are 2 long-living Level1 branches, *develop* and *pnext*? Who's merging stuff between them?

For those of us that need to jump back and forth between these two contexts, its quite a pain. I've spent hours lately just doing context swtiching and trying to get things up and running. Thats just not very inspiring, and very much a waste of resource IMO.

@alemoi, the stuff I'm experiencing in #3029, you thing thats related to this?

**Comment by per-samuelsson** [comment]
> @alemoi, the stuff I'm experiencing in #3029, you thing thats related to this?

Looks like it isn't after all, its something w/ my env most likely: https://github.com/Starcounter/Starcounter/issues/3029#issuecomment-155008391

**Comment by dan31** [comment]
> What strategy do we have now when there are 2 long-living Level1 branches, develop and pnext? Who's merging stuff between them?

@per-samuelsson That one defined by @eriohl is what you look for https://github.com/Starcounter/CompanyTrack/issues/120#issuecomment-151806293 (merge changes from develop into pnext), with an assumption that the merge actually manual and no single person is responsible for the merging process. This actually 1-to-1 to what we've been doing with "next" branch in its days. Please everyone take care of carrying your level1 changes from develop to pnext (CC @chrhol, @alemoi). Once pnext is matured, it will become develop. Once @alemoi is back, we will wrap pnext with more scaffolding tests and apps of our partners, so that the problems we observe will be bumped by both our tests and their code.

**Comment by alemoi** [comment]
Great you brought this topic. Which is an off-topic here now though :smile_cat: 
I myself bumped on the merging policy but in Level0: since I didn't work on `pnext` 3 months ago, I simply didn't merge anything there from `develop`, which was a mistake causing you know what :)
About the automatic tests - I totally agree. Let's define what is suppose to work in `pnext` and move the corresponding tests from `develop`.

**Comment by per-samuelsson** [comment]
> Please everyone take care of carrying your level1 changes from develop to pnext

@dan31, thanks for the clarification, it all very clear inedeed now. Good.

And yeah, really off topic, and the fault is mine. Sorry.

**Comment by mtornwall** [comment]
Apologies for not getting back to you on this sooner. The log output below is from a failed start on the verbose build. Hope it helps. @alemoi 

```
[16:56:23, Notice, Starcounter (networkgateway)]
After ProcessArgumentsAndInitLog

[16:56:23, Notice, Starcounter (networkgateway)]
After AssertCorrectState

[16:56:23, Notice, Starcounter (networkgateway)]
After LoadSettings

[16:56:23, Notice, Starcounter (networkgateway)]
After CreateListeningSocketAndBindToPorts

[16:56:23, Notice, Starcounter (networkgateway)]
After obtaining AcceptEx, ConnectEx, DisconnectEx

[16:56:23, Notice, Starcounter (networkgateway)]
After HttpGlobalInit

[16:56:23, Notice, Starcounter (networkgateway)]
After codegen_uri_matcher_->Init();

[16:56:23, Notice, Starcounter (networkgateway)]
After getting clang functions

[16:56:23, Notice, Starcounter (networkgateway)]
After clangCompileCodeAndGetFuntions_

[16:56:23, Notice, Starcounter (networkgateway)]
After calling clang test functions

[16:56:23, Notice, Starcounter (networkgateway)]
After opening shm_monitor_interface_

[16:56:23, Notice, Starcounter (networkgateway)]
After RegisterGatewayHandlers

[16:56:23, Notice, Starcounter (networkgateway)]
After Init

[16:57:33, Notice, Starcounter (networkgateway)]
After ProcessArgumentsAndInitLog

[16:57:33, Notice, Starcounter (networkgateway)]
After AssertCorrectState

[16:57:33, Notice, Starcounter (networkgateway)]
After LoadSettings

[16:57:33, Notice, Starcounter (networkgateway)]
After CreateListeningSocketAndBindToPorts

[16:57:33, Notice, Starcounter (networkgateway)]
After obtaining AcceptEx, ConnectEx, DisconnectEx

[16:57:33, Notice, Starcounter (networkgateway)]
After HttpGlobalInit

[16:57:34, Notice, Starcounter (networkgateway)]
After codegen_uri_matcher_->Init();

[16:57:35, Notice, Starcounter (networkgateway)]
After getting clang functions

[16:57:38, Error, Starcounter (service)]
ScErrStartNetworkGateway (SCERR2110): Problems starting network gateway process.
```

**Comment by per-samuelsson** [comment]
Are we getting somewhere on this? What's our guess? What does that last log sequence indicate, @alemoi?

**Comment by dan31** [comment]
@per-samuelsson, @alemoi is really busy with https://github.com/Starcounter/CompanyTrack/issues/174, but once he's done this one is a next one to look! @alemoi, confirm :cyclone:

**Comment by alemoi** [comment]
Thanks for reminding! I looked at the log and the following problem is there: LLVM takes too much time to build test code during gateway startup (probably hangs). The reason for this its really hard to say because code there is linear and very simple on our side meaning a possibly yet another issue with LLVM :(

**Comment by alemoi** [comment]
I have produced a test exe for @dan31 to test this specific scenario. He is going to do some testing. @mtornwall here is a file for you as well, I have sent the link by email. Just unpack, run the EXE many times as long as it says Success! and prints small time - its good. If it hangs, then it means that something is wrong with LLVM.

**Comment by alemoi** [comment]
After the discussion we concluded that it also could be some super slowness during startup. For this I included the updated version of scservicelib (that has now 4 times longer waiting time), just replace all `scservicelib*` files in your StarcounterBin (I would suggest also to make a backup first of those files you are going to replace). I also included helper loop.bat to run the test 10000 times, just see if you see any hanging like for several seconds.

**Comment by dan31** [comment]
@alemoi @mtornwall any new insights there?

**Comment by alemoi** [comment]
We came to the conclusion that the problem was with extremely slow startup for scnetworkgateway (just general system slowness, nothing specific to gateway). So what I did is I extended the waiting time for the gateway to start. However, it would be nice to hear the test results from @mtornwall

**Comment by mtornwall** [comment]
As far as LLVM goes, it doesn't seem to be the problem. LLVM initialization time is consistently below 20ms across thousands of runs.

Unfortunately, overlaying the patched scservicelib files on my v2.1.177 installation somehow broke it and it will no longer start.

However, I have now been able to reproduce the startup issue on every machine I've tried it on, including Windows 7, Windows Server 2012 R2, Windows Server 2016 CTP and Windows 10. The problem simply can't be specific to me. :-)
